### PR TITLE
Remove opening transport in 'prepare_for_submission'

### LINKED ISSUE #37: Remove opening transport during `prepare_for_submission`
Currently, during the `prepare_for_submission` we open a transport to inspect the files available on the remote, and then perform logic to decide what to do:

https://github.com/aiidateam/aiida-wannier90/blob/68c03557465b5e0e44d95c5e92549175460abd06/aiida_wannier90/calculations.py#L142-L149

However, this is not ideal, because

- we risk to open too many connections (we skip the part of AiiDA where we pool connections)
- we risk to fail if the connection does not work (we skip the exponential backoff retry)

While we discuss if this can be changed in AiiDA itself (see aiidateam/aiida-core#3477), I think we can apply the following different design: instead of inspecting the files and deciding what to do, we should instead ask explicitly the user where the files are (e.g. via specific flags in the `settings` input node, and maybe require these only if we pass both a remotedata input *and* a folder data input)?
Anyway, for single simulations, the user can check; and for workflows, the workflow probably knows what the previous step retrieved.

### LINKED ISSUE #76: Mutually exclusive remote_input_folder and local_input_folder
Related to #69, #37, #14.

Currently, the logic for selecting input files needs to get the list of files in the `RemoteData`. This opens an SSH transport -- if that fails, the calculation will except, because there is no retry mechanism. In addition, if there are many calculation the remote might block connections. Unlike the regular transport, SSH connections are not re-used. As far as I can tell, there is no easy resolution for this.

A way to avoid having to know the content in the `remote_input_folder` in the first place would be allowing only _either_ `local_input_folder` _or_ `remote_input_folder`, not both at the same time. I discussed this with @giovannipizzi and @AntimoMarrazzo, but we are not 100% sure that all current use cases are covered by this.

@normarivano @qiaojunfeng is this compatible with your current workflows / use cases?

### PR DESCRIPTION
**Pull request by greschd**
Fixes #37 and fixes #76

To avoid opening a transport to the ``remote_input_folder``, we use globbing to specify all _possible_ input files that can be given. Because we can no longer determine if there are overlaps between
local and remote input, the two inputs are now mutually exclusive. Also, the check for the required input files is performed only when a local input is passed. Because passing an input does not make sense in the `postproc_setup` mode, this is no longer allowed.

The implementation separates the specification of input files (which are optional / required, should be copied / symlinked) - which is shared between local and remote input - from the generation of the input lists.

TODO: The `remote_input_folder` version was tested with local transport, but should probably also be tested with ssh transport. It's a somewhat unfortunate consequence of the split between input testing and parser testing that it can't check if the syntax of the input lists is correct (only that it didn't change).

### CODE REVIEW

**Comment by greschd** [comment]
I also took the opportunity to slightly refactor the `prepare_for_submission` - validation of the input and output names was moved to a separate function.

**Comment by greschd** [comment]
@giovannipizzi I'm still tweaking this a little bit.

**Comment by giovannipizzi** [comment]
Quick question - do you want to keep both self.inputs.local_input_folder and self.inputs.remote_input_folder in the end, or you are going to change that part in a different PR?
(see #76)

**Comment by greschd** [comment]
I would keep them both, to keep a path open for implementing mixed input in a backwards-compatible way should it become necessary.

**Comment by greschd** [comment]
Good to go, if the tests pass.

File: aiida_wannier90/calculations.py [hunk 1]
```diff
@@ -257,10 +242,27 @@ def prepare_for_submission(self, folder):  #pylint: disable=too-many-locals, too
...
+                raise exc.InputValidationError(
+                    "None of the 'local_input_folder' and 'remote_input_folder' "
+                    "inputs is set. Exactly one of the two must be given."
+                )
+            if not (has_local_input or has_remote_input):
+                raise exc.InputValidationError(
+                    "Both the 'local_input_folder' and 'remote_input_folder' "
+                    "inputs are set, but they are exclusive. Exactly one of "
+                    "the two must be given."
+                )
```
Comment (giovannipizzi) [inline aiida_wannier90/calculations.py]: I think the two error messages are swapped

File: aiida_wannier90/calculations.py [hunk 2]
```diff
@@ -466,3 +411,119 @@ def _validate_input_parameters(self, parameters):
...
+        return self._get_local_input_file_lists(
+            input_file_specs=input_file_specs,
+            optional_file_globs=optional_file_globs
+        )
+
+    def _get_remote_input_file_lists(
+        self, input_file_specs, optional_file_globs
+    ):
+        """
+        Generate the lists of input files for the case of a remote input folder.
```
Comment (giovannipizzi) [inline aiida_wannier90/calculations.py]: Maybe we should add a comment here explaining why for remote input files we use blobbing (i.e., that in AiiDA if we don't use globbing and the file is not there in the remote, AiiDA will raise an error) - maybe with a link to an appropriate issue on aiida-core?

**Comment by giovannipizzi** [comment]
Note that there might be a small conflict with some changes done in some PRs just merged (on the example for GaAs) - up to you if you want to rebase or merge and fix the confict (if any)

File: aiida_wannier90/calculations.py [hunk 3]
```diff
@@ -466,3 +411,119 @@ def _validate_input_parameters(self, parameters):
...
+                remote_input_folder_uuid,
+                os.path.join(remote_input_folder_path, glob_pattern), '.'
+            )
+
+        remote_copy_list = []
+        remote_symlink_list = [
+            _get_remote_file_info(glob) for glob in optional_file_globs
+        ]
+        for file_spec in input_file_specs:
+            glob = '*' + file_spec.suffix
```
Comment (greschd) [inline aiida_wannier90/calculations.py]: Note to self: for the required files we can use the seedname - we _want_ it to fail if the file is not there.

**Comment by greschd** [comment]
> Note that there might be a small conflict with some changes done in some PRs just merged (on the example for GaAs) - up to you if you want to rebase or merge and fix the confict (if any)

Yeah I had seen that on your PR - we just both fixed the same typo. Nothing complicated, so I'll just merge.

File: aiida_wannier90/calculations.py (hunk 2 above)
Comment (greschd) [reply aiida_wannier90/calculations.py]: Done!

File: aiida_wannier90/calculations.py (hunk 1 above)
Comment (greschd) [reply aiida_wannier90/calculations.py]: Fixed.

**Comment by greschd** [comment]
Ok, I've fixed it according to your comments. Note also that it's now using the _explicit_ file name (instead of globbing) for the _required_ input files -- in that case, it's actually desired that it fails when the file is not there.

**Pull request merged**